cornell-dti/course-plan
# Setup global data store (1/n)

### Summary 

This PR setups the global data store using Vuex. Right now I moved `toggleableRequirementChoices`, `subjectColors`, `uniqueIncrementer` into the store. As a result, we are able to remove various very ugly calls like

`this.$parent.$parent.$parent.createAppCourseFromCornellRosterCourse(...)`

since all the data necessary for creating a new course is available globally, so we don't need to rely on the top-level Dashboard anymore.

In addition, I setup `toggleableRequirementChoices`, `subjectColors`, `uniqueIncrementer` subscribers, so now the realtime firestore data is the source of truth. We no longer need to update them via the toplevel Dashboard anymore. We can simply call the functions to update firestore directly, and the Vuex store and firestore subscriber will take care of the rest.

I moved around some functions into new ts files mostly to avoid cyclic imports.

### Test Plan 

- change your major to A&S biological science, and play with the general chemistry toggleable requirement to verify toggleable requirement still works.
- add a course with a new subject you never added, verify that the new color is generated without issue.
- add a course and verify that the ID is still unique
